generators/server/templates/src/main/java/package/client/TokenRelayRequestInterceptor.java.ejs

```diff
@@ -29,10 +29,18 @@ public class TokenRelayRequestInterceptor implements RequestInterceptor {
 
     public static final String AUTHORIZATION = "Authorization";
 
+    private final AuthorizationHeaderUtil authorizationHeaderUtil;
+
+    public TokenRelayRequestInterceptor(AuthorizationHeaderUtil authorizationHeaderUtil) {
+        super();
+        this.authorizationHeaderUtil = authorizationHeaderUtil;
+    }
+
     @Override
     public void apply(RequestTemplate template) {
-        if (AuthorizationHeaderUtil.getAuthorizationHeader().isPresent()) {
-            template.header(AUTHORIZATION, AuthorizationHeaderUtil.getAuthorizationHeader().get());
+        Optional<String> authorizationHeader = authorizationHeaderUtil.getAuthorizationHeaderFromOAuth2Context();
```

Comment:
In fact, in my repository (https://github.com/farrault/jhipster_oauth2_microservice/pull/1/files) the strategy is different in gateway and microservices applicationType : 
* In a gateway : extract the token from OAuth2Context (and refresh access_token if needed)
    (But for me, a gateway should probably never use it, as most/all call to microservices is done by zuul code)
    via AuthorizationHeaderUtil.getAuthorizationHeaderFrom**OAuth2**Context
* In a microservice : extract the token from OAuth2Authentication (no refresh possible as the refresh token isn't propagated)
    via AuthorizationHeaderUtil.getAuthorizationHeaderFrom**Security**Context

(sorry that this was not very clear in my repo)

As said, previously org.springframework.cloud.security.oauth2.client.feign.OAuth2FeignRequestInterceptor could probably replaced this. However it seems to always use the OAuth2Context .... I thought it wasn't set in a ressource server (the microservices) : this should be checked.

Reply:
In OAuth2 Microservices : 
Based on the source of Spring Boot here : org.springframework.boot.autoconfigure.security.oauth2.client.OAuth2RestOperationsConfiguration.RequestScopedConfiguration
There is indeed a OAuth2ClientContext correctly setup on the request scope that is associated with the access_token received by the microservice

=> That would mean that indeed AuthorizationHeaderUtil.getAuthorizationHeaderFromSecurityContext can be suppressed and the token extracted from OAuth2ClientContext in every case (gateway or microservice) with AuthorizationHeaderUtil.getAuthorizationHeaderFromOAuth2Context

**BUT :** com.mycompany.oauth2gateway.security.oauth2.AuthorizationHeaderUtil.getAuthorizationHeaderFromOAuth2Context()  use a OAuth2RestTemplate to get the OAuth2ClientContext : in a microservice, there is no @Configuration that defines it ( in the current pull request, it is defined by OAuth2SsoConfiguration )
=> Microservice configuration just lack a OAuth2RestTemplate  to make this works : Spring won't start if it doesn't find it

Reply:
Hi, @Tcharl 

For me, a lot more is missing currently for these differents scenarios to work out of the box :
The purpose of this PR is not to add any capability to the current "oauth2" generation (even if there is side effect with the definition of the OAuth2RestTemplate)

In particular, contrary to the "uaa" authenticationType : there is no provided support for client_credentials access_token usage.

For me, it only supports : 
* gateway to microservices via zuul router
  (via `org.springframework.cloud.security.oauth2.proxy.OAuth2TokenRelayFilter` )
  with user access_token refreshing it if necessary
* gateway to microservices via direct feign call (useful ?)
  ( via a feign client annoted with @AuthorizedFeignClient )
  with user access_token refreshing it if necessary (via the custom AuthorizationHeaderUtil : but this complexity may not be justified)
* microservice to microservices via feign
  ( via a feign client annoted with @AuthorizedFeignClient )
  with user access_token WITHOUT being able to refresh it

And the side effects now :
* gateway to another oauth2 app
  via the defined OAuth2RestTemplate
  with user access_token refreshing it if necessary
  __This OAuth2RestTemplate was introduced to allow the refresh in the zuul router case__
* microservice to another oauth2 app
  via the defined OAuth2RestTemplate
  with user access_token WITHOUT being able to refresh it
  __Please note, that this OAuth2RestTemplate is present because I was lazy : I didn't take the time to make AuthorizationHeaderUtil able to work directly with OAuth2ClientContext if there is no OAuth2RestTemplate (or to generate a different implementation in microservice applicationType)__ => however, this OAuth2RestTemplate may be valuable for the usecase you defined

I would suggest to create a dedicated issue about the handling of your different scenarios with oauth2. With the goal to make them work the same way in "oauth2" et "uaa" authenticationType I guess.